File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    gifted = {}  # 각 친구가 누구에게 몇 번의 선물을 주었는지 저장
    gift_idx = {}  # 각 친구의 선물 지수 (받은 선물의 수 - 준 선물의 수)
    
    # 초기화
    for friend in friends:
        gifted[friend] = {}
        gift_idx[friend] = 0
    
    # 선물 기록 처리
    for gift in gifts:
        t, f = gift.split(' ')
        if f in gifted[t]:
            gifted[t][f] += 1
        else:
            gifted[t][f] = 1
        
        gift_idx[t] += 1
        gift_idx[f] -= 1
    
    # 각 친구가 몇 번의 선물을 받을 수 있는지 계산
    will_get = [0 for _ in friends]
    for i in range(len(friends)):
        curr = friends[i]
        for j in range(i + 1, len(friends)):
            another = friends[j]
            
            a = gifted[curr][another] if another in gifted[curr] else 0
            b = gifted[another][curr] if curr in gifted[another] else 0
            
            if a > b:
                will_get[i] += 1
            elif a < b:
                will_get[j] += 1
            elif a == b:
                ai, bi = gift_idx[curr], gift_idx[another]
                if ai > bi:
                    will_get[i] += 1
                elif ai < bi:
                    will_get[j] += 1
    
    answer = max(will_get)
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int n = friends.length;

        Map<String, Integer> idxMap = new HashMap<>();
        for (int i = 0; i < n; i++) {
            idxMap.put(friends[i], i);
        }

        int[][] giftMap = new int[n][n];
        int[] given = new int[n];   
        int[] received = new int[n]; 

        for (String g : gifts) {
            String[] parts = g.split(" ");
            int from = idxMap.get(parts[0]);
            int to = idxMap.get(parts[1]);
            giftMap[from][to]++;
            given[from]++;
            received[to]++;
        }

        int[] nextMonthGifts = new int[n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) continue;

                int giToj = giftMap[i][j];
                int gjToi = giftMap[j][i];

                if (giToj > gjToi) {
                    nextMonthGifts[i]++;
                } else if (giToj == gjToi) {
                    int giScore = given[i] - received[i];
                    int gjScore = given[j] - received[j];
                    if (giScore > gjScore) {
                        nextMonthGifts[i]++;
                    }
                }
            }
        }

        int max = 0;
        for (int x : nextMonthGifts) {
            max = Math.max(max, x);
        }
        return max;
    }
}